### PR TITLE
change extensions.Scale to autoscaling.Scale for apps/v1 statefulset

### LINKED ISSUE #49504: kubectl scale implementation for core workload controllers
This issue covers the proposal for the implementation of kubectl scale for the core controllers.
This is relevant to StatefulSet, ReplicaSet and Deployment.

Currently, kubectl uses [GET/PUT on the resource](https://github.com/kubernetes/kubernetes/blob/c272630b1bc87838cccedd3c04769d2b37b4dc66/pkg/kubectl/scale.go#L268-L277) to update scale, and should instead use the scale endpoint.
The scale endpoint exists in RS and Deployment, and will be added to SS as well shortly.

In order to keep compatibility with older server versions, the proposed change to the `kubectl scale` implementation is as follows:

1. Use the discovery API to check for the presence of a scale sub-resource
2. If scale sub-resource is present, use it for updating the number of replicas
3. If it is absent, fall back to the existing method of GET/PUT on the resource

cc @liggitt @kubernetes/sig-cli-misc @kubernetes/sig-apps-misc

### PR DESCRIPTION
**Pull request by crimsonfaith91**
**What this PR does / why we need it**:
This PR changes `extensions.Scale` to `autoscaling.Scale` for apps/v1 deployment, replicaset, and statefulset.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #49504

**Special notes for your reviewer**:
`autoscaling.Scale` will be used instead of `extensions.Scale` starting apps/v1 group version.

**Release note**:

```release-note
Deployment, ReplicaSet, and StatefulSet uses autoscaling.Scale starting apps/v1 group version.
```

### CODE REVIEW

File: pkg/registry/apps/statefulset/storage/storage.go [hunk 1]
```diff
@@ -174,45 +193,83 @@ func (r *ScaleREST) Update(ctx genericapirequest.Context, name string, objInfo r
 	if obj == nil {
 		return nil, false, errors.NewBadRequest(fmt.Sprintf("nil update passed to Scale"))
 	}
-	scale, ok := obj.(*extensions.Scale)
+
+	var autoScale *autoscaling.Scale
+	var extensionsScale *extensions.Scale
+	switch groupVersion {
```
Comment (crimsonfaith91) [inline pkg/registry/apps/statefulset/storage/storage.go]: Is there a better way than handling both different versions separately?

File: pkg/registry/apps/statefulset/storage/storage_test.go [hunk 2]
```diff
@@ -255,43 +292,80 @@ func TestScaleUpdate(t *testing.T) {
...
+			}
+		}
+
+		obj, err := storage.Scale.Get(ctx, name, &metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("error fetching scale for %s: %v", name, err)
+		}
+
+		switch version {
+		case "v1beta1", "v1beta2":
```
Comment (crimsonfaith91) [inline pkg/registry/apps/statefulset/storage/storage_test.go]: There is a lot of redundant codes due to handling different group versions separately. An idea is to create a generic helper function. I wish to get first round of review before working on the generic helper function. This is to ensure I am on the right track and facilitate code review.

**Comment by crimsonfaith91** [comment]
As first step, I changed `extensions.Scale` to `autoscaling.Scale` for statefulset only. I wish to get first round of code review to ensure I am on the right track before continue working on deployment and replicaset. Same changes are needed for deployment and replicaset, so I expect it will not take a long time to send out a new commit addressing them.

**Comment by liggitt** [comment]
we already have examples of internal type of extensions.Scale accepting/returning external types of apps/v1beta1.Scale (see Convert_v1beta2_Scale_To_extensions_Scale)

Do the same thing by defining conversions for extensions.Scale <-> autoscaling/v1.Scale, and that should be all that is required

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/kubernetes/pull/55413 as an example of switching the internal type to autoscaling.Scale without making any v1-related changes

**Comment by crimsonfaith91** [comment]
@liggitt Thank you very much for the guidance! Appreciate it!

**Comment by crimsonfaith91** [comment]
@liggitt I think the round-tripping test failed due to different `selector` type for `autoscaling.Scale` and `extensions.Scale`. The `selector` field is a string for `autoscaling.Scale`, but it is `LabelSelector` type for `extensions.Scale`.

**Comment by crimsonfaith91** [comment]
Closing in dup of #55413.
Thank you @liggitt for working on it!

**Pull request closed without merging**